students/tim_lurvey/lesson09/mailroom_oo/donor_classes.py

```diff
@@ -0,0 +1,146 @@
+#!/usr/bin/env python
+__author__ = 'Timothy Lurvey'
+
+from running_total import RunningTotal
+
+
+class Donor(RunningTotal):
+    """change attribute from 'key' to 'name'"""
+
+    def __init__(self, name: str, total: float = 0., count: int = 0):
+        super().__init__(new_key=name, total=total, count=count)
+
+    def __repr__(self):
+        return f"Donor('{self.name}', {self.total}, {self.count})"
+
+    @property
+    def name(self):
+        return self.key
+
+    @name.setter
+    def name(self, new_name: str):
+        self.key = new_name
+
+class DonorRepository(object):
+    """Create and manage a data repository of Donor objects.  Donor objects must have unique names."""
+    _REPORT_HEADER = "\nDonor Name                | Total Given | Num Gifts | Average Gift\n" + ("-" * 66) + "\n"
+
+    def __init__(self, data_set: tuple = ()):
+        self._data = []
+        self._set_data(data_set)
+
+    @property
+    def name_list(self) -> tuple:
+        """return a list of all donor names"""
+        return tuple(sorted([d.name for d in self._data]))
+
+    def _expand_input_values(self, data: (list, tuple, set)) -> tuple:
+        """expand all possible values from sequence"""
+        n = data[0]
+        t = 0.
+        c = 0
+        try:
+            t += float(data[1])
+            c += int(data[2])
+        except:
+            pass
+        return n, t, c
+
+    def _set_data(self, data_set) -> None:
+        """work through the inputs to create a populated data object"""
+        for d_obj in data_set:
+            if isinstance(d_obj, Donor):
+                add_obj = d_obj
+            else:
+                n, t, c = self._expand_input_values(d_obj)
+                add_obj = Donor(name=n, total=t, count=c)
+            self._data.append(add_obj)
+
+    def add_new_donor(self, obj) -> None:
+        """add a Donor object or a sequence of data to be converted (name, total, count)
+        Donor('x', 1. ,1 )or ('x', 102.5, 2)
+        In the data input sequence, total and count are optional"""
+        try:
+            assert isinstance(obj, Donor) or isinstance(obj, (list, tuple, set))
```

Comment:
assert statements should not be used in "production" code

Reply:
What is the most pythonic way to type check objects and raise an error?  I have previously used a "If not isinstance(...): raise TypeError(...)" but the If ... else, to me, is more for flow control, not for type checking.

Reply:
yes if/raise is the proper way to handle error checking in production code